components/motor/ulnstepper/28byj-48.go

```diff
@@ -0,0 +1,478 @@
+			actualBoard, motorConfig, err := getBoardFromRobotConfig(deps, config)
+			if err != nil {
+				return nil, err
+			}
+
+			return newULN(ctx, actualBoard, *motorConfig, config.Name, logger)
+		},
+	}
+	registry.RegisterComponent(motor.Subtype, model, _motor)
+	config.RegisterComponentAttributeMapConverter(
+		motor.Subtype,
+		model,
+		func(attributes config.AttributeMap) (interface{}, error) {
+			var conf Config
+			return config.TransformAttributeMapToStruct(&conf, attributes)
+		},
+		&Config{},
+	)
+}
+
+func getBoardFromRobotConfig(deps registry.Dependencies, config config.Component) (board.Board, *Config, error) {
+	motorConfig, ok := config.ConvertedAttributes.(*Config)
+	if !ok {
+		return nil, nil, rdkutils.NewUnexpectedTypeError(motorConfig, config.ConvertedAttributes)
+	}
+	if motorConfig.BoardName == "" {
+		return nil, nil, errors.New("expected board name in config for motor")
+	}
+	b, err := board.FromDependencies(deps, motorConfig.BoardName)
+	if err != nil {
+		return nil, nil, err
+	}
+	return b, motorConfig, nil
+}
+
+func newULN(ctx context.Context, b board.Board, mc Config, name string, logger golog.Logger) (motor.Motor, error) {
+	if mc.TicksPerRotation == 0 {
+		return nil, errors.New("expected ticks_per_rotation in config for motor")
+	}
+
+	if mc.RotationPerMinute == 0 {
+		return nil, errors.New("expected rotation_per_minute value greater than zero")
+	}
+
+	m := &uln2003{
+		theBoard:          b,
+		rotationPerMinute: mc.RotationPerMinute,
+		ticksPerRotation:  mc.TicksPerRotation,
+		logger:            logger,
+		motorName:         name,
+	}
+
+	if mc.Pins.In1 != "" {
+		in1, err := b.GPIOPinByName(mc.Pins.In1)
+		if err != nil {
+			return nil, err
+		}
+		m.in1 = in1
+	}
+
+	if mc.Pins.In2 != "" {
+		in2, err := b.GPIOPinByName(mc.Pins.In2)
+		if err != nil {
+			return nil, err
+		}
+		m.in2 = in2
+	}
+
+	if mc.Pins.In3 != "" {
+		in3, err := b.GPIOPinByName(mc.Pins.In3)
+		if err != nil {
+			return nil, err
+		}
+		m.in3 = in3
+	}
+
+	if mc.Pins.In4 != "" {
+		in4, err := b.GPIOPinByName(mc.Pins.In4)
+		if err != nil {
+			return nil, err
+		}
+		m.in4 = in4
+	}
+
+	if err := m.Validate(); err != nil {
+		return nil, err
+	}
+
+	m.startThread(ctx)
+	return m, nil
+}
+
+// struct is named after the controler uln2003.
+type uln2003 struct {
+	theBoard           board.Board
+	rotationPerMinute  float64
+	ticksPerRotation   int
+	in1, in2, in3, in4 board.GPIOPin
+	logger             golog.Logger
+	motorName          string
+
+	// state
+	lock  sync.Mutex
+	opMgr operation.SingleOperationManager
+
+	stepPosition         int64
+	threadStarted        bool
+	targetStepPosition   int64
+	targetStepsPerSecond int64
+	generic.Unimplemented
+}
+
+// validate if this config is valid.
+func (m *uln2003) Validate() error {
+	if m.theBoard == nil {
+		return errors.Errorf("need a board for motor (%s)", m.motorName)
+	}
+
+	if m.rotationPerMinute == 0 {
```

Comment:
[minor] if this isn't a config parameter, then we set a default in `newULN` based on the minimums and use that.